slybot/slybot/plugins/scrapely_annotations/builder.py

```diff
@@ -294,53 +294,84 @@ def apply_selector_annotations(annotations, target_page):
-                chain(*[[elem._root for elem in page.css(sel)]
-                        for sel in annotation.get('reject_selectors', [])
-                        if sel])
-            )
-            elems = accepted_elements - rejected_elements
-        else:
-            elems = [elem._root for elem in page.css(annotation['selector'])]
-        if not elems:
-            continue
-
-        tagids = [int(e.attrib.get('data-tagid', 1e9)) for e in elems]
-        tagid = min(tagids)
+        if annotation.get('item_container'):
+            containers[annotation['id']] = annotation
+        selector = annotation.get('selector')
+        tagid, elems = tagid_for_annotation(annotation, page)
         if tagid is not None:
             annotation['tagid'] = tagid
+            if selector:
+                tagid_selector_map[tagid] = selector
             converted_annotations.append(annotation)
 
         # Create container for repeated field annotation
         if (annotation.get('repeated') and
                 not annotation.get('item_container') and
                 len(annotation.get('annotations')) == 1):
-            parent = _get_parent(elems, page)
-            field = annotation['annotations'].values()[0][0]['field']
-            container_id = '%s#parent' % annotation['id']
-            if len(parent):
-                converted_annotations.append({
-                    'item_container': True,
-                    'id': container_id,
-                    'annotations': {'#portia-content': '#dummy'},
-                    'text-content': '#portia-content',
-                    'container_id': annotation['container_id'],
-                    'field': field,
-                    'tagid': parent.attrib.get('data-tagid')
-                })
-                annotation['item_container'] = True
-                annotation['field'] = field
-                annotation['container_id'] = container_id
+            repeated_parent = add_repeated_field(annotation, elems, page)
+            if repeated_parent:
+                converted_annotations.append(repeated_parent)
+                container_id = repeated_parent['container_id']
+                added_repeated[container_id] = repeated_parent
+    if added_repeated:
+        for container_id, child in added_repeated.items():
+            container = containers[container_id]
+            if container['tagid'] != child['tagid']:
+                continue
+            _, elems = tagid_for_annotation(container, page)
+            parent = elems[0].getparent()
+            container['tagid'] = int(parent.attrib.get('data-tagid', 1e9))
     return converted_annotations
 
 
+def tagid_for_annotation(annotation, page):
+    selector = annotation.get('selector')
+    if not selector:
+        accepted_elements = set(
+            chain(*[[elem._root for elem in page.css(sel)]
+                    for sel in annotation.get('accept_selectors', [])
+                    if sel])
```

Comment:
Refactor these chains? The only things that changes is `accept_selectors` vs `reject_selectors`